### ISSUE TITLE
log4js pulls in security vulnerabilities

### ISSUE THREAD

**Issue by kolbma** [log4js-node/log4js-node#706]
via mailgun-js

```
┌────────────┬────────────────────────────────────────────────────────────────────┐
│            │ Denial of Service                                                  │
├────────────┼────────────────────────────────────────────────────────────────────┤
│ Name       │ http-proxy-agent                                                   │
├────────────┼────────────────────────────────────────────────────────────────────┤
│ CVSS       │ 8 (High)                                                           │
├────────────┼────────────────────────────────────────────────────────────────────┤
│ Installed  │ 1.0.0                                                              │
├────────────┼────────────────────────────────────────────────────────────────────┤
│ Vulnerable │ <=2.0.0                                                            │
├────────────┼────────────────────────────────────────────────────────────────────┤
│ Patched    │ >=2.1.0                                                            │
├────────────┼────────────────────────────────────────────────────────────────────┤
│ Path       │ nodejs-intro-testing-sample@0.0.1 > karma@2.0.2 > log4js@2.5.3 >   │
│            │ mailgun-js@0.7.15 > proxy-agent@2.0.0 > http-proxy-agent@1.0.0     │
├────────────┼────────────────────────────────────────────────────────────────────┤
│ More Info  │ https://nodesecurity.io/advisories/607                             │
└────────────┴────────────────────────────────────────────────────────────────────┘

┌────────────┬────────────────────────────────────────────────────────────────────┐
│            │ Denial of Service                                                  │
├────────────┼────────────────────────────────────────────────────────────────────┤
│ Name       │ https-proxy-agent                                                  │
├────────────┼────────────────────────────────────────────────────────────────────┤
│ CVSS       │ 8 (High)                                                           │
├────────────┼────────────────────────────────────────────────────────────────────┤
│ Installed  │ 1.0.0                                                              │
├────────────┼────────────────────────────────────────────────────────────────────┤
│ Vulnerable │ <=2.1.1                                                            │
├────────────┼────────────────────────────────────────────────────────────────────┤
│ Patched    │ >=2.2.0                                                            │
├────────────┼────────────────────────────────────────────────────────────────────┤
│ Path       │ nodejs-intro-testing-sample@0.0.1 > karma@2.0.2 > log4js@2.5.3 >   │
│            │ mailgun-js@0.7.15 > proxy-agent@2.0.0 > https-proxy-agent@1.0.0    │
├────────────┼────────────────────────────────────────────────────────────────────┤
│ More Info  │ https://nodesecurity.io/advisories/593                             │
└────────────┴────────────────────────────────────────────────────────────────────┘

┌────────────┬────────────────────────────────────────────────────────────────────┐
│            │ Regular Expression Denial of Service                               │
├────────────┼────────────────────────────────────────────────────────────────────┤
│ Name       │ timespan                                                           │
├────────────┼────────────────────────────────────────────────────────────────────┤
│ CVSS       │ 3.7 (Low)                                                          │
├────────────┼────────────────────────────────────────────────────────────────────┤
│ Installed  │ 2.3.0                                                              │
├────────────┼────────────────────────────────────────────────────────────────────┤
│ Vulnerable │ All                                                                │
├────────────┼────────────────────────────────────────────────────────────────────┤
│ Patched    │ None                                                               │
├────────────┼────────────────────────────────────────────────────────────────────┤
│ Path       │ nodejs-intro-testing-sample@0.0.1 > karma@2.0.2 > log4js@2.5.3 >   │
│            │ loggly@1.1.1 > timespan@2.3.0                                      │
├────────────┼────────────────────────────────────────────────────────────────────┤
│ More Info  │ https://nodesecurity.io/advisories/533                             │
└────────────┴────────────────────────────────────────────────────────────────────┘

┌────────────┬────────────────────────────────────────────────────────────────────┐
│            │ Regular Expression Denial of Service                               │
├────────────┼────────────────────────────────────────────────────────────────────┤
│ Name       │ debug                                                              │
├────────────┼────────────────────────────────────────────────────────────────────┤
│ CVSS       │ 3.7 (Low)                                                          │
├────────────┼────────────────────────────────────────────────────────────────────┤
│ Installed  │ 2.2.0                                                              │
├────────────┼────────────────────────────────────────────────────────────────────┤
│ Vulnerable │ <= 2.6.8 || >= 3.0.0 <= 3.0.1                                      │
├────────────┼────────────────────────────────────────────────────────────────────┤
│ Patched    │ >= 2.6.9 < 3.0.0 || >= 3.1.0                                       │
├────────────┼────────────────────────────────────────────────────────────────────┤
│ Path       │ nodejs-intro-testing-sample@0.0.1 > karma@2.0.2 > log4js@2.5.3 >   │
│            │ mailgun-js@0.7.15 > debug@2.2.0                                    │
├────────────┼────────────────────────────────────────────────────────────────────┤
│ More Info  │ https://nodesecurity.io/advisories/534                             │
└────────────┴────────────────────────────────────────────────────────────────────┘
```

**Comment by nomiddlename** [comment]
These are optional dependencies. If you do not use the mailgun appender, the code will not be loaded. If you do use the mailgun appender, then there is an updated appender available: `npm i @log4js-node/mailgun`. Same goes for the loggly appender. These appender will be removed from the core in version 3.x

**Comment by kolbma** [comment]
Good to know.  
But I don't think this is a good solution to depend on an optional dependency to remove a vulnerability from required dependencies.
  
Projects doesn't depend only on _log4js_ package and currently I've got shown 90 vulnerabilities by `npm audit`. If package maintainers argument all the same... it is optional, you can update yourself, the way we use the dependency it is not vulnerable and so on, the `npm audit` or `nsp` and more important the `npm update` is good-for-nothing.

If I run `npm update` on my project there shouldn't be any known vulnerabilities left after a period of time. And with period of time I don't think of more than 6 months because of the next major release.

Then in this case I've to wait for an update of karma package to your 3.0 version because of `^2.5.3`.

Are there so difficult incompatibilities in mailgun, that you can't update the version and release a version __2.5.4__ with little effort? That's the way vulnerabilities should get handled. 

Sorry, that I sound abrasively, but I'm shocked by 180 vulnerabilities after `npm update && npm audit`.  
That's definitively not the way to not get laughed down by customers if suggesting a node.js solution.

**Comment by nicojs** [comment]
@nomiddlename 

> These appender will be removed from the core in version 3.x

Do you mean the appender will be removed from `"log4js"`@3.x? If so, what is the time table on that? I want to update from 1.x to 2.x, but with these audit messages, it seems that it is better to wait for 3.x.

**Comment by nomiddlename** [comment]
I want to get all the optional appenders deprecated before I release 3.x, I'm about halfway through that. I can't give you a timetable, because I only have a couple of hours a week to spend on this and I'm doing it for free.

I'll take a look at updating the mailgun dependency, and see if it breaks anything.

**Comment by nomiddlename** [comment]
Published to npm in v2.6.0, updated mailgun-js to 0.18.0. Let me know if this doesn't fix the problem.

**Comment by iamjoyce** [comment]
After running `npm audit` and on v2.6.1, I am still having 6 vulnerabilities from `log4js`.

Please let me know if I should create a new issue.

```
  moderate        Prototype pollution

  Package         hoek

  Dependency of   log4js

  Path            log4js > loggly > request > hawk > boom > hoek

  More info       https://nodesecurity.io/advisories/566


  moderate        Prototype pollution

  Package         hoek

  Dependency of   log4js

  Path            log4js > loggly > request > hawk > cryptiles > boom > hoek

  More info       https://nodesecurity.io/advisories/566


  moderate        Prototype pollution

  Package         hoek

  Dependency of   log4js

  Path            log4js > loggly > request > hawk > hoek

  More info       https://nodesecurity.io/advisories/566


  moderate        Prototype pollution

  Package         hoek

  Dependency of   log4js

  Path            log4js > loggly > request > hawk > sntp > hoek

  More info       https://nodesecurity.io/advisories/566


  moderate        Memory Exposure

  Package         tunnel-agent

  Dependency of   log4js

  Path            log4js > loggly > request > tunnel-agent

  More info       https://nodesecurity.io/advisories/598


  low             Regular Expression Denial of Service

  Package         timespan

  Dependency of   log4js

  Path            log4js > loggly > timespan

  More info       https://nodesecurity.io/advisories/533
```
